kernel/gic/src/gic/mod.rs

```diff
@@ -23,38 +23,100 @@ pub type InterruptNumber = u32;
 /// 8-bit unsigned integer
 pub type Priority = u8;
 
-bitflags! {
-    /// The legacy (GICv2) way of specifying
-    /// multiple target CPUs
-    pub struct TargetList: u8 {
-        const CPU_0 = 1 << 0;
-        const CPU_1 = 1 << 1;
-        const CPU_2 = 1 << 2;
-        const CPU_3 = 1 << 3;
-        const CPU_4 = 1 << 4;
-        const CPU_5 = 1 << 5;
-        const CPU_6 = 1 << 6;
-        const CPU_7 = 1 << 7;
-        const ALL_CPUS = u8::MAX;
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct TargetList(u8);
+
+impl TargetList {
+    pub fn new(list: &[CpuId]) -> Result<Self, &'static str> {
+        let mut this = 0;
+
+        for cpu_id in list {
+            let mpidr = MpidrValue::from(*cpu_id).value();
+            if mpidr < 8 {
+                this |= 1 << mpidr;
+            } else {
+                return Err("CPU id is too big for GICv2 (should be < 8)");
+            }
+        }
+
+        Ok(Self(this))
+    }
+
+    /// Tries to create a `TargetList` from `arm_boards::BOARD_CONFIG.cpu_ids`
+    pub fn new_all_cpus() -> Result<Self, &'static str> {
+        let cpu_ids: [CpuId; NUM_CPUS] = core::array::from_fn(|i| {
+            CpuId::from(BOARD_CONFIG.cpu_ids[i])
+        });
+
+        Self::new(&cpu_ids).map_err(|_| "Some CPUs in the system cannot be stored in a TargetList")
```

Comment:
if `Self::new()` accepts an iterator, you won't need to try to create a fixed array here.